File: Solutions/r/AdventOfCode.r
get_date <- function(args) {
    date <- strsplit(args[1], split = "/")
    return(date)
}

get_solution <- function(year, day) {
    result <- paste("Solutions/r/y", year, "/Solution", day, ".r", sep = "")
    return(result)
}

get_inputfile <- function(year, day) {
    result <- paste("Inputs/", year, "/", day, ".txt", sep = "")
    return(result)
}

solve <- function(solution, input_file) {
    source(solution)
    part1(input_file)  # nolint
    part2(input_file)  # nolint
}

args <- commandArgs(trailingOnly = TRUE)
date <- get_date(args)
year <- date[[1]][1]
day <- date[[1]][2]
input_file <- get_inputfile(year, day)
solution <- get_solution(year, day)

solve(solution, input_file)

File: Solutions/lua/AdventOfCode.lua
local util = require("Solutions.lua.util")

local function getdate(arg)
    local dateStr = arg[1]
    local date = util.split(dateStr, "/")
    return date
end

local function getsolution(date)
    local solutionName = "Solutions.lua.y" .. date[1] .. ".solution" .. date[2]
    return solutionName
end

local function getinputfile(date)
    local inputfile = "Inputs/" .. date[1] .. "/" .. date[2] .. ".txt"
    return inputfile
end

local function solve(solutionName, inputfile)
    local solution = require(solutionName)
    solution.part1(inputfile)
    solution.part2(inputfile)
end

local date = getdate(arg)
local solutionName = getsolution(date)
local inputfile = arg[2]

if (inputfile == nil) then
    inputfile = getinputfile(date)
end

solve(solutionName, inputfile)

